// @mobile-reality/react-native-select-pro/dist/module/components/multi-select/index.js
import React from 'react';
import { ScrollView, StyleSheet, useWindowDimensions } from 'react-native';
import { parsePercentageToNumber } from '../../helpers';
import { SelectInput } from '../select-input';
import { MultiSelectedOption } from './components/selected-option';
export const MultiSelect = _ref => {
  let {
    searchable,
    selectControlTextStyle,
    selectControlStyle,
    selectedOption,
    placeholderText,
    placeholderTextColor,
    onPressRemove,
    disabled,
    dispatch,
    isOpened,
    onPressSelectControl,
    searchPattern,
    searchValue,
    setPosition,
    multiSelection,
    multiSelectionOptionStyle
  } = _ref;
  const {
    width
  } = useWindowDimensions();
  const selectedOptionTyped = selectedOption;

  const resolveSelectedOptions = () => {
    if (!selectedOptionTyped) {
      if (searchable) {
        return null;
      }

      return /*#__PURE__*/React.createElement(MultiSelectedOption, {
        isPlaceholder: true,
        option: null,
        optionWidth: "100%",
        placeholderText: placeholderText,
        placeholderTextColor: placeholderTextColor,
        selectControlTextStyle: selectControlTextStyle
      });
    }

    const optionWidth = () => {
      const WIDTH_THRESHOLD = 100;
      const WIDTH_OFFSET = 72;
      const {
        length
      } = selectedOptionTyped;
      const initialWidth = selectControlStyle ? selectControlStyle.width : 100;
      let calculatedWidth = 100;

      if (typeof initialWidth === 'number') {
        calculatedWidth = (initialWidth - WIDTH_OFFSET) / length;

        if (calculatedWidth < WIDTH_THRESHOLD) {
          return WIDTH_THRESHOLD;
        }

        return Math.floor(calculatedWidth);
      }

      if (typeof initialWidth === 'string') {
        const ratioToScreen = Math.floor(width * (parsePercentageToNumber(initialWidth) / 100));
        calculatedWidth = ratioToScreen / length;

        if (calculatedWidth - WIDTH_OFFSET < WIDTH_THRESHOLD) {
          return WIDTH_THRESHOLD;
        }

        return calculatedWidth - WIDTH_OFFSET;
      }

      return 0;
    };

    return selectedOptionTyped.map((option, index) => {
      return /*#__PURE__*/React.createElement(MultiSelectedOption, {
        key: index,
        option: option,
        optionWidth: optionWidth(),
        placeholderText: placeholderText,
        selectControlTextStyle: selectControlTextStyle,
        multiSelectionOptionStyle: multiSelectionOptionStyle,
        onPressRemove: onPressRemove
      });
    });
  };

  return /*#__PURE__*/React.createElement(ScrollView, {
    horizontal: true,
    style: styles.multiSelectionWrapper
  }, searchable && /*#__PURE__*/React.createElement(SelectInput, {
    disabled: disabled,
    dispatch: dispatch,
    isOpened: isOpened,
    multiSelection: multiSelection,
    placeholderText: placeholderText,
    placeholderTextColor: placeholderTextColor,
    searchPattern: searchPattern,
    searchValue: searchValue,
    selectControlTextStyle: selectControlTextStyle,
    selectedOption: selectedOption,
    setPosition: setPosition,
    onPressSelectControl: onPressSelectControl
  }), resolveSelectedOptions());
};
const styles = StyleSheet.create({
  multiSelectionWrapper: {
    flex: 1
  }
});
