// @mobile-reality/react-native-select-pro/dist/module/components/clear-option/index.js
import React from 'react';
import { Image, Pressable, StyleSheet } from 'react-native';
export const ClearOption = _ref => {
  let {
    selectControlClearOptionA11yLabel,
    disabled,
    selectControlClearOptionButtonHitSlop,
    selectControlClearOptionButtonStyle,
    selectControlClearOptionImageStyle,
    onPressRemove
  } = _ref;
  return /*#__PURE__*/React.createElement(Pressable, {
    accessibilityLabel: selectControlClearOptionA11yLabel || 'Clear a chosen option',
    accessibilityRole: "button",
    accessible: true,
    disabled: disabled,
    hitSlop: selectControlClearOptionButtonHitSlop ? selectControlClearOptionButtonHitSlop : {
      right: 3,
      left: 3
    },
    style: [styles.xIconWrapper, selectControlClearOptionButtonStyle],
    onPress: onPressRemove
  }, /*#__PURE__*/React.createElement(Image, {
    source: require('./../../assets/icons/x.png'),
    style: [styles.xIcon, selectControlClearOptionImageStyle]
  }));
};
const styles = StyleSheet.create({
  xIconWrapper: {
    height: '100%',
    justifyContent: 'center'
  },
  xIcon: {
    width: 20,
    height: 20,
    zIndex: 1
  }
});
