// @mobile-reality/react-native-select-pro/dist/commonjs/components/multi-select/components/selected-option/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultiSelectedOption = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _styles = require("../../../../constants/styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MultiSelectedOption = _ref => {
  var _StyleSheet$flatten;

  let {
    isPlaceholder = false,
    option,
    selectControlTextStyle,
    multiSelectionOptionStyle,
    placeholderText,
    placeholderTextColor,
    optionWidth,
    onPressRemove
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_reactNative.Pressable, {
    accessibilityLabel: option ? `${option.label} selected` : 'Placeholder in multi-select',
    style: [isPlaceholder ? styles.placeholderText : styles.multiSelectionOption, multiSelectionOptionStyle, {
      width: optionWidth
    }],
    onPress: () => onPressRemove ? onPressRemove(option) : undefined
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    numberOfLines: 1,
    style: [styles.text, selectControlTextStyle, {
      color: option !== null && option !== void 0 && option.label ? ((_StyleSheet$flatten = _reactNative.StyleSheet.flatten(selectControlTextStyle)) === null || _StyleSheet$flatten === void 0 ? void 0 : _StyleSheet$flatten.color) || _styles.COLORS.BLACK : placeholderTextColor
    }]
  }, (option === null || option === void 0 ? void 0 : option.label) || placeholderText));
};

exports.MultiSelectedOption = MultiSelectedOption;

const styles = _reactNative.StyleSheet.create({
  text: {
    fontSize: _styles.FONT_SIZE,
    textAlign: 'left'
  },
  placeholderText: {
    alignItems: 'flex-start',
    justifyContent: 'center',
    display: 'flex'
  },
  multiSelectionOption: {
    alignItems: 'flex-start',
    justifyContent: 'center',
    display: 'flex',
    borderRadius: 4,
    backgroundColor: _styles.COLORS.DISABLED,
    borderWidth: 1,
    borderColor: _styles.COLORS.BLACK,
    margin: 2,
    paddingLeft: 5
  }
});
