// @mobile-reality/react-native-select-pro/dist/commonjs/components/clear-option/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClearOption = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const ClearOption = _ref => {
  let {
    selectControlClearOptionA11yLabel,
    disabled,
    selectControlClearOptionButtonHitSlop,
    selectControlClearOptionButtonStyle,
    selectControlClearOptionImageStyle,
    onPressRemove
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_reactNative.Pressable, {
    accessibilityLabel: selectControlClearOptionA11yLabel || 'Clear a chosen option',
    accessibilityRole: "button",
    accessible: true,
    disabled: disabled,
    hitSlop: selectControlClearOptionButtonHitSlop ? selectControlClearOptionButtonHitSlop : {
      right: 3,
      left: 3
    },
    style: [styles.xIconWrapper, selectControlClearOptionButtonStyle],
    onPress: onPressRemove
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: require('./../../assets/icons/x.png'),
    style: [styles.xIcon, selectControlClearOptionImageStyle]
  }));
};

exports.ClearOption = ClearOption;

const styles = _reactNative.StyleSheet.create({
  xIconWrapper: {
    height: '100%',
    justifyContent: 'center'
  },
  xIcon: {
    width: 20,
    height: 20,
    zIndex: 1
  }
});
