// @mobile-reality/react-native-select-pro/dist/commonjs/components/select-control/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectControl = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = require("../../constants/styles");

var _helpers = require("../../helpers");

var _types = require("../../state/types");

var _clearOption = require("../clear-option");

var _multiSelect = require("../multi-select");

var _selectInput = require("../select-input");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const arrowImage = require('./../../assets/icons/chevron-down.png');

const SelectControl = /*#__PURE__*/(0, _react.forwardRef)((_ref, ref) => {
  let {
    isOpened,
    animated,
    animationDuration,
    selectControlStyle,
    selectedOption,
    onPressSelectControl,
    dispatch,
    clearable,
    options,
    disabled,
    multiSelection,
    placeholderText,
    placeholderTextColor,
    searchable,
    searchPattern,
    searchValue,
    setPosition,
    selectControlArrowImageStyle,
    selectControlDisabledStyle,
    selectControlClearOptionButtonHitSlop,
    selectControlClearOptionButtonStyle,
    selectControlClearOptionImageStyle,
    selectControlClearOptionA11yLabel,
    selectControlOpenDropdownA11yLabel,
    selectControlButtonsContainerStyle,
    multiSelectionOptionStyle,
    hideSelectControlArrow,
    onSelect,
    onRemove,
    selectControlTextStyle,
    aboveSelectControl,
    customLeftIconSource,
    customLeftIconStyles,
    selectedOptionIndex
  } = _ref;
  const rotateAnimation = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
  (0, _react.useEffect)(() => {
    if (animated) {
      _reactNative.Animated.timing(rotateAnimation, {
        toValue: isOpened ? 1 : 0,
        duration: animationDuration,
        useNativeDriver: true
      }).start();
    }
  }, [rotateAnimation, isOpened, animated]);
  const rotate = rotateAnimation.interpolate({
    inputRange: [0, 1],
    outputRange: ['0deg', '180deg']
  });

  const onPressRemove = function () {
    let option = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    if (!disabled) {
      let removedOption = selectedOption;
      let removedOptionIndex = selectedOptionIndex;

      if (multiSelection) {
        let removedSelectedOptions = [];
        removedSelectedOptions = selectedOption.filter(selected => selected.value !== option.value);

        if (removedSelectedOptions.length === 0) {
          removedSelectedOptions = null;
        }

        const foundIndex = options.findIndex(_ref2 => {
          let {
            value
          } = _ref2;
          return value === (option === null || option === void 0 ? void 0 : option.value);
        });
        removedOptionIndex = foundIndex;
        removedOption = option;
        const resolveSelectedOptionIndex = selectedOptionIndex.filter(item => item !== foundIndex);
        dispatch({
          type: _types.Action.SelectOption,
          payload: {
            selectedOption: removedSelectedOptions,
            selectedOptionIndex: (resolveSelectedOptionIndex === null || resolveSelectedOptionIndex === void 0 ? void 0 : resolveSelectedOptionIndex.length) > 0 ? resolveSelectedOptionIndex : -1
          }
        });
      } else {
        dispatch({
          type: _types.Action.SelectOption,
          payload: {
            selectedOption: null,
            selectedOptionIndex: -1
          }
        });

        if (searchable) {
          dispatch({
            type: _types.Action.SetSearchValue,
            payload: ''
          });
        }

        dispatch({
          type: _types.Action.SetOptionsData,
          payload: options
        });
      }

      if (onSelect) {
        onSelect(null, -1);
      }

      if (onRemove) {
        onRemove(removedOption, removedOptionIndex);
      }
    }
  };

  const [isScreenReaderEnabled, setIsScreenReaderEnabled] = (0, _react.useState)(false);
  (0, _react.useEffect)(() => {
    if (!_helpers.isAndroid) {
      _reactNative.AccessibilityInfo.isScreenReaderEnabled().then(e => {
        setIsScreenReaderEnabled(e);
      });

      _reactNative.AccessibilityInfo.addEventListener('change', e => {
        setIsScreenReaderEnabled(e);
      });
    }
  }, []);
  const isShowClearOptionButton = clearable && selectedOption && !isScreenReaderEnabled;
  const isShowClearOptionButtonA11y = clearable && selectedOption && isScreenReaderEnabled && !_helpers.isAndroid;

  const renderArrowImage = () => {
    const accessibilityLabel = 'Arrow for opening dropdown';
    const arrow = animated ? /*#__PURE__*/_react.default.createElement(_reactNative.Animated.Image, {
      source: arrowImage,
      style: [styles.arrowIcon, {
        transform: [{
          rotate
        }]
      }, selectControlArrowImageStyle]
    }) : /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
      source: arrowImage,
      style: [styles.arrowIcon, isOpened ? styles.arrowIconOpened : styles.arrowIconClosed, selectControlArrowImageStyle]
    });

    if (multiSelection) {
      return /*#__PURE__*/_react.default.createElement(_reactNative.Pressable, {
        accessibilityLabel: accessibilityLabel,
        onPress: disabled ? undefined : onPressSelectControl
      }, arrow);
    }

    return arrow;
  };

  const renderMultiselect = () => {
    return /*#__PURE__*/_react.default.createElement(_multiSelect.MultiSelect, {
      disabled: disabled,
      dispatch: dispatch,
      isOpened: isOpened,
      multiSelection: multiSelection,
      placeholderText: placeholderText,
      placeholderTextColor: placeholderTextColor,
      searchPattern: searchPattern,
      searchValue: searchValue,
      searchable: searchable,
      selectControlStyle: selectControlStyle,
      selectControlTextStyle: selectControlTextStyle,
      multiSelectionOptionStyle: multiSelectionOptionStyle,
      selectedOption: selectedOption,
      setPosition: setPosition,
      onPressRemove: onPressRemove,
      onPressSelectControl: onPressSelectControl
    });
  };

  const renderSelection = () => {
    var _StyleSheet$flatten;

    const selectedOptionTyped = selectedOption; // for proper typing

    if (searchable) {
      return /*#__PURE__*/_react.default.createElement(_selectInput.SelectInput, {
        disabled: disabled,
        dispatch: dispatch,
        isOpened: isOpened,
        multiSelection: multiSelection,
        placeholderText: placeholderText,
        placeholderTextColor: placeholderTextColor,
        searchPattern: searchPattern,
        searchValue: searchValue,
        selectControlTextStyle: selectControlTextStyle,
        selectedOption: selectedOption,
        setPosition: setPosition,
        onPressSelectControl: onPressSelectControl
      });
    }

    return /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
      numberOfLines: 1,
      style: [styles.text, selectControlTextStyle, {
        color: selectedOptionTyped !== null && selectedOptionTyped !== void 0 && selectedOptionTyped.label ? ((_StyleSheet$flatten = _reactNative.StyleSheet.flatten(selectControlTextStyle)) === null || _StyleSheet$flatten === void 0 ? void 0 : _StyleSheet$flatten.color) || _styles.COLORS.BLACK : placeholderTextColor
      }]
    }, (selectedOptionTyped === null || selectedOptionTyped === void 0 ? void 0 : selectedOptionTyped.label) || placeholderText);
  };

  const resolveAccessibilityHint = () => {
    if (!selectedOption) {
      return undefined;
    }

    if (!multiSelection) {
      const selectedOptionTyped = selectedOption; // for proper typing

      return `Current selected item is ${selectedOptionTyped === null || selectedOptionTyped === void 0 ? void 0 : selectedOptionTyped.label}`;
    }

    return 'You have selected multiple items';
  };

  const resolveContainer = () => {
    if (multiSelection && selectedOption) {
      return {
        Component: _reactNative.View
      };
    }

    return {
      Component: _reactNative.Pressable
    };
  };

  const {
    Component
  } = resolveContainer();
  const shouldRenderClearButton = isShowClearOptionButton && !multiSelection;
  const shouldRenderClearButtonA11y = isShowClearOptionButtonA11y && !multiSelection;
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.rootView
  }, /*#__PURE__*/_react.default.createElement(Component, {
    ref: ref,
    accessibilityHint: resolveAccessibilityHint(),
    accessibilityLabel: isOpened ? '' : selectControlOpenDropdownA11yLabel || 'Open a dropdown',
    style: [styles.container, isOpened ? aboveSelectControl ? styles.openedAbove : styles.opened : {}, selectControlStyle, disabled ? [styles.disabled, selectControlDisabledStyle] : {}],
    onPress: disabled || multiSelection && selectedOption ? undefined : onPressSelectControl
  }, !!customLeftIconSource && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.leftIconWrapper, styles.xIconWrapper]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: customLeftIconSource,
    style: customLeftIconStyles
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.press, multiSelection ? styles.pressMultiSelection : styles.pressNormal]
  }, multiSelection ? renderMultiselect() : renderSelection()), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.iconsContainer, selectControlButtonsContainerStyle]
  }, shouldRenderClearButton && /*#__PURE__*/_react.default.createElement(_clearOption.ClearOption, {
    disabled: disabled,
    selectControlClearOptionA11yLabel: selectControlClearOptionA11yLabel,
    selectControlClearOptionButtonHitSlop: selectControlClearOptionButtonHitSlop,
    selectControlClearOptionButtonStyle: selectControlClearOptionButtonStyle,
    selectControlClearOptionImageStyle: selectControlClearOptionImageStyle,
    onPressRemove: onPressRemove
  }), !hideSelectControlArrow && renderArrowImage())), shouldRenderClearButtonA11y && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.a11IconWrapper
  }, /*#__PURE__*/_react.default.createElement(_clearOption.ClearOption, {
    disabled: disabled,
    selectControlClearOptionA11yLabel: selectControlClearOptionA11yLabel,
    selectControlClearOptionButtonHitSlop: selectControlClearOptionButtonHitSlop,
    selectControlClearOptionButtonStyle: selectControlClearOptionButtonStyle,
    selectControlClearOptionImageStyle: selectControlClearOptionImageStyle,
    onPressRemove: onPressRemove
  })));
});
exports.SelectControl = SelectControl;

const styles = _reactNative.StyleSheet.create({
  rootView: {
    position: 'relative'
  },
  container: {
    height: 40,
    flexDirection: 'row',
    borderRadius: _styles.SHAPE,
    borderWidth: _styles.BORDER_WIDTH,
    backgroundColor: _styles.COLORS.WHITE
  },
  press: {
    flex: 1,
    height: '100%',
    paddingHorizontal: _styles.PADDING,
    justifyContent: 'center'
  },
  pressMultiSelection: {
    paddingRight: 40
  },
  pressNormal: {
    paddingRight: 55
  },
  disabled: {
    backgroundColor: _styles.COLORS.DISABLED
  },
  text: {
    fontSize: _styles.FONT_SIZE,
    textAlign: 'left'
  },
  openedAbove: {
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0
  },
  opened: {
    borderBottomLeftRadius: 0,
    borderBottomRightRadius: 0
  },
  iconsContainer: {
    position: 'absolute',
    right: 8,
    top: 0,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    height: '100%'
  },
  arrowIcon: {
    width: 25,
    height: 25,
    zIndex: -1
  },
  leftIconWrapper: {
    paddingLeft: 8
  },
  xIconWrapper: {
    height: '100%',
    justifyContent: 'center'
  },
  arrowIconOpened: {
    transform: [{
      rotate: '180deg'
    }]
  },
  arrowIconClosed: {
    transform: [{
      rotate: '0deg'
    }]
  },
  a11IconWrapper: {
    position: 'absolute',
    right: -20,
    borderWidth: 1,
    height: '100%'
  }
});

SelectControl.displayName = 'SelectControl';
