// @mobile-reality/react-native-select-pro/dist/module/components/option/index.js
import React, { forwardRef } from 'react';
import { StyleSheet, Text, TouchableOpacity } from 'react-native';
import { COLORS, FONT_SIZE, ITEM_HEIGHT, PADDING } from '../../constants/styles';
export const Option = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    optionSelectedStyle,
    optionStyle,
    optionTextStyle,
    isSelected,
    onPressOption,
    option,
    onSelect,
    OptionComponent,
    optionIndex
  } = _ref;
  const {
    label
  } = option;

  const onChooseOption = () => {
    onPressOption(option, optionIndex);

    if (onSelect) {
      onSelect(option, optionIndex);
    }
  };

  if (OptionComponent) {
    return /*#__PURE__*/React.createElement(OptionComponent, {
      isSelected: isSelected,
      option: option,
      onPressOption: onChooseOption
    });
  }

  return /*#__PURE__*/React.createElement(TouchableOpacity, {
    ref: ref,
    accessibilityLabel: `Choose ${label} option`,
    accessibilityRole: "button",
    accessible: true,
    style: [styles.option, optionStyle, isSelected && [styles.selected, optionSelectedStyle]],
    onPress: onChooseOption
  }, /*#__PURE__*/React.createElement(Text, {
    numberOfLines: 1,
    style: [styles.text, optionTextStyle]
  }, label));
});
const styles = StyleSheet.create({
  option: {
    height: ITEM_HEIGHT,
    justifyContent: 'center',
    paddingHorizontal: PADDING
  },
  text: {
    fontSize: FONT_SIZE,
    color: COLORS.BLACK,
    textAlign: 'left'
  },
  selected: {
    backgroundColor: COLORS.SELECTED
  }
});
Option.displayName = 'Option';
